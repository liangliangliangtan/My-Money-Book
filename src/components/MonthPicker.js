/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react'
import PropTypes from 'prop-types'
import { padZeroToLeft, range} from '../utilities/PriceUtility';

/**
 * Month Picker Component 
 * 
 * 1. Create a button to display years and months from props 
 * 2. Create dropdown menu, onClick event call a callback function to hide or show
 *    dropdown List
 * 3. In drowpdown List to show month list and year list
 * 
 * 4. to add hightlight to those selected year and month (className = " *** active")
 * 5. Interactive function when you select year and month
 *     select year : change the selected year state and updates 
 *     select month : close the dropdown list and invoke the callback function with 
 *                   parameters selected year and selected month 
 * 
 * 6. Optimize:  click other area to close the drop down list
 *      
 *      
 */
class MonthPicker extends React.Component{

    constructor(props){
        super(props)
        this.state = { 
            isOpenDropDown: false,
            selectedYear: this.props.year,
            selectedMonth: this.props.month
        }
        this.stopPropagation = this.stopPropagation.bind(this);
        
    }

    /**
     * Add listener when MonthPicker has constructed 
     * Listen all the click event on this page 
     *  
     */
    componentDidMount() {
        document.addEventListener('click', this.hideAndSubmit)
    }

    componentWillUnmount(){
        document.removeEventListener('click',this.hideAndSubmit)
    }
    
    /**
     * When the user clicks other area, 
     * close the dropDown list and invoke the callback function
     */
    hideAndSubmit = (event) => {
        //event.preventDefault();
        this.setState({
            isOpenDropDown: false
        })
        this.props.onChange(this.state.selectedYear, this.state.selectedMonth)
    }

    toggleDropdown = (event) =>{
        this.stopPropagation(event);
        event.preventDefault();
        this.setState({
            isOpenDropDown: !this.state.isOpenDropDown
        })
    }

    generateHighLight = (current, target) =>{
        return current === target ? "dropdown-item active" :  "dropdown-item"
    }

    setSelectedYear = (event, yearNumber) =>{
        this.stopPropagation(event);
        event.preventDefault();
        this.setState({
            selectedYear: yearNumber
        })
    }

    /**
     * need to tell the parent component that year and month has been changed 
     */
    setSelectedMonth = (event, monthNumber) =>{
        this.stopPropagation(event);
        event.preventDefault();
        this.setState({
            isOpenDropDown:false,
            selectedMonth: monthNumber
        })
        this.props.onChange(this.state.selectedYear, monthNumber)
    }

    /**
     * Use Event.stopImmediatePropagation to prevent your listeners on the root from being called. 
     * It is supported in IE9+ and modern browsers.
     * @param {*} event 
     */
    stopPropagation(event) {
        event.nativeEvent.stopImmediatePropagation();
    }
    

    render(){
        const { isOpenDropDown, selectedYear, selectedMonth} = this.state
        const monthRange = range(12,1) // 1 ~ 12
        const yearRange = range(9,-4).map(number => number + this.props.year) // (-4 ~ +4) + current year
        return (
            <div className="month-picker-component">
                <div className="dropdown">
                    <p>Please select month and year</p>
                    <button className="btn btn-lg btn-dark dropdown-toggle text-justify" id="dropdownMenuButton"
                        data-toggle="dropdown" aria-haspopup="true" aria-expanded="true"
                        onClick={(event) => this.toggleDropdown(event)}
                        >
                        {this.state.selectedYear} / {padZeroToLeft(this.state.selectedMonth)}
                    </button>
                    {isOpenDropDown &&
                        <div className=" dropdown-menu text-justify" 
                             aria-labelledby="dropdownMenuButton"
                             style={{ display: 'block', alignSelf: 'center', margin: '0 auto', left:'40%'}} 
                             role="menu">
                            <div className="row">
                                <div className="col border-right year-ranges">
                                    {yearRange.map((yearNumber, index) =>(
                                        <a key={index} className={this.generateHighLight(yearNumber, selectedYear)} href="#" 
                                            onClick={(event) => this.setSelectedYear(event,yearNumber)}>
                                            {yearNumber} 
                                        </a>
                                    ))}
                                </div>
                                <div className="col month-ranges">
                                    {monthRange.map((monthNumber,index) => (
                                        <a key={index} className={this.generateHighLight(monthNumber, selectedMonth)} href ="#"
                                            onClick={(event) => this.setSelectedMonth(event, monthNumber)}>
                                            {padZeroToLeft(monthNumber)}
                                        </a>
                                    ))}
                                </div>
                            </div>
                        </div>
                    }
                    
                </div>
            </div>
        )
    }
}

MonthPicker.Prototype = {
    year: PropTypes.number.isRequired,
    month: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired
}

export default MonthPicker;