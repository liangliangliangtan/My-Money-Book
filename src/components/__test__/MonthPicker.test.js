import React from 'react'
import {mount} from 'enzyme'
import MonthPicker from '../MonthPicker'
import toJson from 'enzyme-to-json';
import { noop } from '@babel/types';


/**
 * Mount:(https://gist.github.com/fokusferit/e4558d384e4e9cab95d04e5f35d4f913)
    The only way to test componentDidMount and componentDidUpdate. 
    Full rendering including child components.
    Requires a DOM (jsdom, domino). 
    More constly in execution time.
 */

let props = { 
    year: 2019,
    month: 6,
    onChange: jest.fn()
}

let nativeEvent = { nativeEvent: { stopImmediatePropagation: noop } }

let wrapper
describe('test Month Picker Component',() =>{

    beforeEach(()=>{
        wrapper = mount(<MonthPicker {...props}/>)
    })

    it('should render the component match the snapshot',()=>{
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('render the correct year and month,show the correct dropdown state', () =>{
        let text = wrapper.find('.dropdown-toggle').text() // the year and month shown on the button 
        expect(text).toEqual('2019 / 06')
        expect(wrapper.find('.dropdown-menu').length).toEqual(0)
        expect(wrapper.state('isOpenDropDown')).toEqual(false)
        expect(wrapper.state('selectedYear')).toEqual(props.year)
        expect(wrapper.state('selectedMonth')).toEqual(props.month)
    })
    
    it('after click the button, the drop-down should show, and yearlist & monthlist should have correct length',()=>{
        //mock click event
        wrapper.find('.dropdown-toggle').simulate('click', nativeEvent)
        expect(wrapper.state('isOpenDropDown')).toEqual(true)
        expect(wrapper.find('.dropdown-menu').length).toEqual(1)
        // year list has length 9 and month list has lenght 12
        expect(wrapper.find('.year-ranges a').length).toEqual(9)
        expect(wrapper.find('.month-ranges a').length).toEqual(12)
        // hightlight text should be  2019 year, 06 month(a tag with className= 'dropdown-item active')
        expect(wrapper.find('.year-ranges .dropdown-item.active').text()).toEqual('2019')
        expect(wrapper.find('.month-ranges .dropdown-item.active').text()).toEqual('06')
        // first year shown on the dropdown menu should be equal to 2019 - 4 = 2015 
        expect(wrapper.find('.year-ranges a').first().text()).toEqual(`${props.year-4}`)
    })

    it('click the year and month item should trigger the right status change',()=>{
        // click the first year, the selected  year state should be changed, and its 
        // a tag should be with active className 
        wrapper.find('.dropdown-toggle').simulate('click', nativeEvent)
        wrapper.find('.year-ranges a').first().simulate('click', nativeEvent)
        expect(wrapper.find('.year-ranges a').first().hasClass('active')).toEqual(true)
        expect(wrapper.state('selectedYear')).toEqual(props.year-4)
        // click the last month(December) 
        // 1.the state of seleceted month should be changed
        // 2.the props should be called with 2015,1
        // 3.thedrop down  items should be closed 
        wrapper.find('.month-ranges a').last().simulate('click', nativeEvent)
        expect(wrapper.state('isOpenDropDown')).toEqual(false)
        expect(wrapper.state('selectedMonth')).toEqual(12)
        expect(props.onChange).toHaveBeenCalledWith(2015,12)
    })

    it('If the dropdwon is shown, click the document should close the dropdown',()=>{
        // mock document.addEventListener
        let eventMap = {}
        document.addEventListener = jest.fn((event,cb) =>{
            eventMap[event] = cb 
        })
        // create the addEventListener at construction time 
        const wrapper = mount(<MonthPicker {...props} />)
        wrapper.find('.dropdown-toggle').simulate('click', nativeEvent)
        expect(wrapper.state('isOpenDropDown')).toEqual(true)
        expect(wrapper.find('.dropdown-menu').length).toEqual(1)
        // if click the document , the drop down menu should be closed
        eventMap.click({
            target: document
        })
        expect(wrapper.state('isOpenDropDown')).toEqual(false)
    })
})