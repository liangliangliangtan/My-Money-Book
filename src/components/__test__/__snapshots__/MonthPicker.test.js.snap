// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Month Picker Component should render the component match the snapshot 1`] = `
<MonthPicker
  month={6}
  onChange={[MockFunction]}
  year={2019}
>
  <div
    className="dropdown pull-right month-picker-component"
  >
    <h5>
      Please select month and year
    </h5>
    <button
      className="btn btn-lg btn-dark dropdown-toggle"
      onClick={[Function]}
    >
      2019
       / 
      06
      <span
        className="caret"
      />
    </button>
  </div>
</MonthPicker>
`;
